### PR TITLE
PR updated all pdb files to pattern files (done)

### PR DESCRIPTION
**Pull request by cramoisi**
1. updating .gitignore with https://github.com/koreader/crengine/pull/206#issuecomment-396819150

2. TODO : 
~'Bulgarian_hyphen_(Alan).pdb'~
~'Hungarian_hyphen_(Alan).pdb'~
~'Portuguese_hyphen_(Alan).pdb'~
~'Russian_hyphen_(Alan).pdb'~
~'Ukrain_hyphen_(Alan).pdb'~
~'Czech_hyphen_(Alan).pdb'~
~'Icelandic_hyphen_(Alan).pdb'~
~'Roman_hyphen_(Alan).pdb'~
~'Slovak_hyphen_(Alan).pdb'~
~'Danish_hyphen_(Alan).pdb'~
~'Irish_hyphen_(Alan).pdb'~
~'Russian_EnGB_hyphen_(Alan).pdb'~
~'Slovenian_hyphen_(Alan).pdb'~
~'Finnish_hyphen_(Alan).pdb'~
~'Italian_hyphen_(Alan).pdb'~
~'Russian_EnUS_hyphen_(Alan).pdb'~
~'Swedish_hyphen_(Alan).pdb'~

### CODE REVIEW

**Comment by poire-z** [comment]
Can you add `cr3gui/data/dict_ext/` too ? (not really documented feature, but a 2nd lookup is done with this dir only, and I have it on my emulator)

**Comment by cramoisi** [comment]
Yup

**Comment by cramoisi** [comment]
Let's hold the merge until all pdb files are converted to pattern files :-)

**Comment by cramoisi** [comment]
@poire-z RT , I don't like the idea (and can't test it before night) but what would happen if we keep the same filename.ext with an xml inside ?

**Comment by poire-z** [comment]
Looks like the filename matters, and the extension is used to decide which code to use to read that file content:
https://github.com/koreader/crengine/blob/9e2cb90edb580688e97f42c3a3397702dc2081bd/crengine/src/hyphman.cpp#L298-L306

**Comment by poire-z** [comment]
It won't hurt keeping these lines (in case our only Luxembourgeois or Serb user made its own Luxembourgeois.pdb or Serbian.pdb :)

**Comment by cramoisi** [comment]
(ok, I got my procedure right for each langage) : json / pattern / delete or it fails

**Comment by cramoisi** [comment]
@poire-z : in a ideal world, I could add all languages supported [there](https://www.ushuaia.pl/hyphen/?ln=en) or use in libreoffice

**Comment by cramoisi** [comment]
ok, need a break

**Comment by cramoisi** [comment]
side note : the italian pattern could probably be improve the same way I improve the french one as it use some patterns with quote and typographic quote which are ignored (and it's a good thing) by crengine. If there is italian words which contains these, they need specific rules.

**Comment by cramoisi** [comment]
and it's done !

File: .gitignore [hunk 1]
```diff
@@ -15,4 +15,5 @@ cr3gui/data/manual/
 cr3gui/data/cr3.css
 cr3gui/data/cr3.ini
 cr3gui/data/dict/
+cr3gui/data/dict_ext/
```
Comment (Frenzie) [inline .gitignore]: What does it do?

Comment (cramoisi) [reply .gitignore]: @Frenzie  : @poire-z request this one : https://github.com/koreader/crengine/pull/209#issuecomment-396858686

**Comment by poire-z** [comment]
No `left_hyphen_min/right_hyphen_min` forgotten with your 2nd batch?

@Frenzie : I added that `dict_ext` (in readerdictionary.lua) for the reason explained in https://github.com/koreader/koreader/issues/2393#issue-192982601 :

> stardict : do a 2nd query using dicts in data/dict_ext/ (if it exists) : I put in dict_ext/ the "English Wiktionary, Western languages" dict, which has definitions for mostly any declination of a word in all western languages. If I put it in dict/, it finds an exact definition that my other french dicts don't have, and masks the sometimes valuable french definition I would have had with fuzzy searches if that Wiktionary wasn't there. With that 2nd lookup, i get both.

**Comment by cramoisi** [comment]
@poire-z : the default value is 2/2 for everything ; i will add them after my morning coffee :)

**Comment by poire-z** [comment]
The defaults, when not specified, are 2/2. So may be you don't need to add them, to not overload that file, and so we can quickly see those that don't use the defaults?

**Comment by poire-z** [comment]
If that ^ is fine, and you're fine with this PR, I'd like to merge it so I can rebase the soft-hyphens PR and avoid conflicts, so you can test it.

**Comment by cramoisi** [comment]
@poire-z  yup, I didn't add them because they were not explicitly done for another config than 2/2 and I wanted your advice. So, it's ok with me, you can merge.

**Comment by cramoisi** [comment]
I was just thinking about that. The fact was the 2/2 was explicitly defined in the file so, I add them. Let's remove them.

**Comment by poire-z** [comment]
The only use of it is there:
https://github.com/koreader/crengine/blob/4f6dd9180563186cd0c9a95771c36a238345c122/crengine/src/lvrend.cpp#L1159-L1165
so, the only value supported is "auto" (vs "none").

But there are 4 hyphenate CSS properties that are supported, with different possible values:
https://github.com/koreader/crengine/blob/4f6dd9180563186cd0c9a95771c36a238345c122/crengine/src/lvstsheet.cpp#L106-L109
https://github.com/koreader/crengine/blob/4f6dd9180563186cd0c9a95771c36a238345c122/crengine/src/lvstsheet.cpp#L662-L684
https://github.com/koreader/crengine/blob/4f6dd9180563186cd0c9a95771c36a238345c122/crengine/src/lvstsheet.cpp#L913-L923

A tweak with `* {hyphenate: auto !important} ` should work, but if not, you can try setting the 4 css properties to auto?

**Comment by cramoisi** [comment]
the auto will make the hyphenation with the pattern file ?

**Comment by poire-z** [comment]
Not exactly: `auto` will make use of the hyphenation manager (and then, the method you chose, pattern, algo or none, will be used). Otherwise, it's like the `none` method is used.

**Comment by poire-z** [comment]
`* {hyphenate: inherit !important}` should work, as the root element is set to `auto`, and hyphenate is inherited. But let's see if the other tweak works.

**Comment by cramoisi** [comment]
I need to re-edit an ebook to test it ; I was fed up and get rid of the css declarations because it was evening and I wanted to read and not to test'n tweak :-)

languages.json is done. I also removed the two default line for english as they were 2/2

**Comment by cramoisi** [comment]
@poire-z : it only strike my mind now but : 

> A tweak with * {hyphenate: auto !important} should work, but if not, you can try setting the 4 css properties to auto?

`hyphenate` don't exist in css, it's : 

```
hyphens:
-webkit-hyphens:
-moz-hyphens:
-ms-hyphens:
-o-hyphens:
adobe-hyphenate:
```

Looks like it's a case of a class name mistaken for a properties name (again)

they all take none | manual | inherit values but the adobe one which takes explicit in place of manual

**Comment by cramoisi** [comment]
but anyway, it doesn't work as intended (all hyph properties are set to none in the publisher css, my do_hyph styletweak sets all these to auto (or inherit) and even that, nothing changes)

**Comment by poire-z** [comment]
Well, to minimize changes (and may be cause `hyphenate:` exists somewhere else like in fb2 files, and it does at least in the obsoleted *.css), may be we should just in
https://github.com/koreader/crengine/blob/4f6dd9180563186cd0c9a95771c36a238345c122/crengine/src/lvstsheet.cpp#L106-L109
rename `-webkit-hyphens:` to `hyphens:` ?

As all these get mapped to a single internal style property (style->hyphenate), any Style tweak with any value would override any of the value set previously from another property (and if used with `!important`, they would not be overridden by embedded styles with any of these property).

Could you try it and confirm it works as expected (and make a PR if it does :) ?

**Comment by cramoisi** [comment]
ok (I get it working). Removing the -moz will break the support for hyphenation enabled only `.kepub.epub` files... But i've yet to see an ebook with only one line for hyphenation in its css. Usually publishers put them all just to be sure (let's check with kobo desktop app and the only kepub i ever bought from there)

**Comment by poire-z** [comment]
If I set in one wikipedia article stylesheet.css: `p { hyphenate: none; }`
When opening it, nothing is hyphenated.
If I use a styletweak with: `p { hyphenate: auto !important; }` and enable it, text gets hyphenated, according to the setting selected in our `Hyphenation>` menu.
So, I guess a style tweak with `* { hyphenate: auto !important; }` should work.

So, whatever the property name is, the following values should do these things:
- `inherit` => **inherit rules are applied**, and if nothing met along, the default of "auto" set to the root element would be used.
- `none` or `optimizeSpeed `=> **none**: words won't be hyphenated at all, no matter what we selected in our `Hyphenation>` menu
- `auto`, `optimizeQuality` or `explicit` => **auto**: words will be hyphenated with the method selected in our `Hyphenation>` menu
- any other value, including the unsupported `manual`  => **invalid, ignored**

I worried for a few seconds there, because I remembed i had seen a lot of `optimizeSpeed` at the time I had to edit embedded stylesheets , but it was in `text-rendering : optimizeSpeed;`, so fortunatly, no impact on hyphenation.

**Comment by cramoisi** [comment]
yes it works as intended, it just my rule that was higher than expected to be right as I set it to * and the publisher set none to p.

>rename -webkit-hyphens: to hyphens: ?

Yup. Nickel do its own stuff so they don't rely on the css to hyphenate (more than that : kepubs works best without anything hyphenation related in the css). ADE use the adobe stuff so, ok, I'll PR a change on -webkit.

**Comment by cramoisi** [comment]
`adobe-text-layout`,  first time i see this. an InDesign thing ? not clean not standard not pretty  :'(

**Comment by poire-z** [comment]
Asked Google, mostly results from 2008-2010, Adobe Text Layout Framework, Flesk, Flash stuff...
Doesn't hurt to keep it, unless it's the one you prefer to sacrifice to make room for `hyphens:`

**Comment by cramoisi** [comment]
> unless it's the one you prefer to sacrifice to make room for hyphens:

A great idea !, yes !

**Comment by poire-z** [comment]
You can re-order them if you feel like it, shouldn't matter:
https://github.com/koreader/crengine/blob/944d90fcc8e2b9c0bd6b17f189c0760e5562a6b2/crengine/src/lvstsheet.cpp#L106-L109
Update the comments too in:
https://github.com/koreader/crengine/blob/944d90fcc8e2b9c0bd6b17f189c0760e5562a6b2/crengine/src/lvstsheet.cpp#L34-L37

**Pull request merged**